source/extensions/transport_sockets/tls/io_handle_bio.cc

```diff
@@ -0,0 +1,140 @@
+#include "extensions/transport_sockets/tls/io_handle_bio.h"
+
+#include "envoy/buffer/buffer.h"
+#include "envoy/network/io_handle.h"
+
+#include "openssl/bio.h"
+#include "openssl/err.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace TransportSockets {
+namespace Tls {
+
+namespace {
+
+// NOLINTNEXTLINE(readability-identifier-naming)
+inline Envoy::Network::IoHandle* bio_io_handle(BIO* bio) {
+  return reinterpret_cast<Envoy::Network::IoHandle*>(bio->ptr);
+}
+
+// NOLINTNEXTLINE(readability-identifier-naming)
+int io_handle_new(BIO* bio) {
+  bio->init = 0;
+  bio->num = -1;
+  bio->ptr = nullptr;
+  bio->flags = 0;
+  return 1;
+}
+
+// NOLINTNEXTLINE(readability-identifier-naming)
+int io_handle_free(BIO* bio) {
+  if (bio == nullptr) {
+    return 0;
+  }
+
+  if (bio->shutdown) {
+    if (bio->init) {
+      bio_io_handle(bio)->close();
+    }
+    bio->init = 0;
+    bio->flags = 0;
+  }
+  return 1;
+}
+
+// NOLINTNEXTLINE(readability-identifier-naming)
+int io_handle_read(BIO* b, char* out, int outl) {
+  if (out == nullptr) {
+    return 0;
+  }
+
+  Envoy::Buffer::RawSlice slice;
+  slice.mem_ = out;
+  slice.len_ = outl;
+  auto result = bio_io_handle(b)->readv(outl, &slice, 1);
+  BIO_clear_retry_flags(b);
+  if (!result.ok()) {
+    if (result.err_->getErrorCode() == Api::IoError::IoErrorCode::Again) {
+      BIO_set_retry_read(b);
+    }
+    return -1;
+  }
+  return result.rc_;
+}
+
+// NOLINTNEXTLINE(readability-identifier-naming)
+int io_handle_write(BIO* b, const char* in, int inl) {
+  Envoy::Buffer::RawSlice slice;
+  slice.mem_ = const_cast<char*>(in);
+  slice.len_ = inl;
+  auto result = bio_io_handle(b)->writev(&slice, 1);
+  BIO_clear_retry_flags(b);
+  if (!result.ok()) {
+    if (result.err_->getErrorCode() == Api::IoError::IoErrorCode::Again) {
```

Comment:
We may also want to set retry in the case of Api::IoError::IoErrorCode::Interrupt.

Although it seems that writes returning Interrupt will result in Envoy closing the socket.  I don't know how often a EINTR return would happen in practice.

Reply:
Interesting. I guess we could treat EINTR as EAGAIN here.